chore: changes in GitHub Actions automation

ankur0904 opened the pull request:
Fixes: #30

Make changes in `package.json` file as suggest in #30 

cc: @derberg

derberg commented:
Thanks a lot, as per the comment in linked issue, please also make changes to https://github.com/asyncapi/EDAVisualiser/blob/master/.github/workflows/github-pages.yml

- `actions/checkout@v2` change to `actions/checkout@v3`
- instead of step below
  ```yml
        - name: Setup Node.js
          uses: actions/setup-node@v2
          with:
            node-version: 14
            cache: 'npm'
            cache-dependency-path: '**/package-lock.json'
  ```
  we need 
  ```yml
        - name: Check package-lock version
          uses: asyncapi/.github/.github/actions/get-node-version-from-package-lock@master
          id: lockversion
        - name: Setup Node.js
          uses: actions/setup-node@v3
          with:
            node-version: "${{ steps.lockversion.outputs.version }}"
            cache: 'npm'
            cache-dependency-path: '**/package-lock.json'
  ```

ankur0904 commented:
@derberg Made the required change.

derberg reviewed (dismissed):
great work! @ankur0904 if you want there are more issues like this

@jonaslagoni good to merge from my side - also new workflows already pushed https://github.com/asyncapi/EDAVisualiser/pull/49

jonaslagoni commented:
/rtm

jonaslagoni commented:
/rtm

derberg commented:
tests fail as installation has issues, something with lock file, not sure what though as you did not change version of lock file with your PR 🤔 

```
npm info run husky@2.7.0 install node_modules/husky node husky install
npm info run husky@2.7.0 install { code: 0, signal: null }

> postinstall
> npm run bootstrap

npm verb cli /opt/hostedtoolcache/node/18.18.0/x64/bin/node /opt/hostedtoolcache/node/18.18.0/x64/bin/npm
npm info using npm@9.8.1
npm info using node@v18.18.0
npm verb title npm run bootstrap
npm verb argv "run" "bootstrap"
npm verb logfile logs-max:10 dir:/home/runner/.npm/_logs/2023-10-11T13_48_19_911Z-
npm verb logfile /home/runner/.npm/_logs/2023-10-11T13_48_19_911Z-debug-0.log

> bootstrap
> lerna bootstrap

lerna notice cli v3.22.1
lerna info ci enabled
lerna info Bootstrapping 1 package
lerna info Installing external dependencies
lerna ERR! npm ci exited 1 in '@asyncapi/edavisualiser'
lerna ERR! npm ci stderr:
npm verb cli /opt/hostedtoolcache/node/18.18.0/x64/bin/node /opt/hostedtoolcache/node/18.18.0/x64/bin/npm
npm info using npm@9.8.1
npm info using node@v18.18.0
npm verb title npm ci
npm verb argv "ci"
npm verb logfile logs-max:10 dir:/home/runner/.npm/_logs/2023-10-11T13_48_21_138Z-
npm verb logfile /home/runner/.npm/_logs/2023-10-11T13_48_21_138Z-debug-0.log
npm WARN old lockfile 
npm WARN old lockfile The package-lock.json file was created with an old version of npm,
npm WARN old lockfile so supplemental metadata must be fetched from the registry.
npm WARN old lockfile 
npm WARN old lockfile This is a one-time fix-up, please be patient...
npm WARN old lockfile 
```

ankur0904 commented:
@derberg I think I mistakely include the `package-lock.json` in the PR.

ankur0904 commented:
@derberg Can I undo the change in `package-lock.json` and make a commit ?

derberg commented:
but you did good that you included it, because you removed some dependencies, and proper info needed to be updated in package lock

try to remove it on local, then run `npm install --verbose` and then commit new version of the package lock. Remember to use node 18 in case of this repo

ankur0904 commented:
@derberg Made the required changes in `package-lock.json`.

jonaslagoni reviewed (commented):
Seems like `jest` is not installed as a dev dependency 🤔 

How it worked before I have no idea 🧐 

Try and install Jest as a dev dependency 🙂

derberg commented:
@jonaslagoni I saw this `jest` installation issue on multiple repos several times, it was always related to a conflict of `node used in the ci` vs `package lock version`

now I see what is the problem, I did not pay enough attention and did not know this repo is a monorepo

@ankur0904 
- remove jest that you added as it is already here https://github.com/asyncapi/EDAVisualiser/blob/master/library/package.json
- root package lock is on version 3 so pipelines use node 18, but what I just noticed that in library package locks in on version 1 (https://github.com/asyncapi/EDAVisualiser/blob/master/library/package-lock.json) so yeah, it is definitely an issue because pipelines use node 18 with latest npm that has issues parsing package-lock version 1. Solution - follow my previous instruction with package lock refresh, but on `library/package-lock.json`

ankur0904 commented:
@derberg @jonaslagoni 
Made the required update.

derberg commented:
I see there are some more installation issues:

```
lerna notice cli v3.22.1
lerna info ci enabled
lerna info Bootstrapping 1 package
lerna info Installing external dependencies
lerna ERR! npm ci exited 1 in '@asyncapi/edavisualiser'
lerna ERR! npm ci stderr:
npm verb cli /opt/hostedtoolcache/node/18.18.0/x64/bin/node /opt/hostedtoolcache/node/18.18.0/x64/bin/npm
npm info using npm@9.8.1
npm info using node@v18.18.0
npm verb title npm ci
npm verb argv "ci"
npm verb logfile logs-max:10 dir:/home/runner/.npm/_logs/2023-10-13T19_10_09_493Z-
npm verb logfile /home/runner/.npm/_logs/2023-10-13T19_10_09_493Z-debug-0.log
npm http fetch GET 200 https://registry.npmjs.org/webpack 254ms (cache miss)
npm verb stack Error: could not resolve
npm verb stack     at PlaceDep.failPeerConflict (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/place-dep.js:503:25)
npm verb stack     at new PlaceDep (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/place-dep.js:157:21)
npm verb stack     at #buildDepStep (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:933:18)
npm verb stack     at async Arborist.buildIdealTree (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:196:7)
npm verb stack     at async CI.exec (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/lib/commands/ci.js:61:5)
npm verb stack     at async module.exports (/opt/hostedtoolcache/node/18.18.0/x64/lib/node_modules/npm/lib/cli-entry.js:61:5)
npm verb cwd /home/runner/work/EDAVisualiser/EDAVisualiser/library
npm verb Linux 6.2.0-1012-azure
npm verb node v18.18.0
npm verb npm  v9.8.1
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: babel-loader@9.1.3
npm ERR! Found: webpack@4.47.0
npm ERR! node_modules/webpack
npm ERR!   dev webpack@"^4.41.2" from the root project
npm ERR!   peer webpack@"^4 || ^5" from @cypress/webpack-preprocessor@5.17.1
npm ERR!   node_modules/@cypress/webpack-preprocessor
npm ERR!     dev @cypress/webpack-preprocessor@"^5.9.0" from the root project
npm ERR!   3 more (terser-webpack-plugin, ts-loader, webpack-cli)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer webpack@">=5" from babel-loader@9.1.3
npm ERR! node_modules/babel-loader
npm ERR!   peer babel-loader@"^8.0.2 || ^9" from @cypress/webpack-preprocessor@5.17.1
npm ERR!   node_modules/@cypress/webpack-preprocessor
npm ERR!     dev @cypress/webpack-preprocessor@"^5.9.0" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: webpack@5.88.2
npm ERR! node_modules/webpack
npm ERR!   peer webpack@">=5" from babel-loader@9.1.3
npm ERR!   node_modules/babel-loader
npm ERR!     peer babel-loader@"^8.0.2 || ^9" from @cypress/webpack-preprocessor@5.17.1
npm ERR!     node_modules/@cypress/webpack-preprocessor
npm ERR!       dev @cypress/webpack-preprocessor@"^5.9.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
```

looks like `webpack` from `library` need a bump. @jonaslagoni you need to help.

@ankur0904 what I can confirm though is that this installation issue can be replicated on local

ankur0904 commented:
@derberg 
**I ran the command**
```bash
npm install --loglevel verbose
```
**But it give the permission error**
```bash
[Error: EACCES: permission denied, open '/home/ankur/Desktop/development/EDAVisualiser/library/styles/default.css'] {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/home/ankur/Desktop/development/EDAVisualiser/library/styles/default.css'
}
npm verb exit 1
npm verb code 1
npm verb exit 1
npm verb code 1
npm verb exit 1
npm verb code 1
lerna info lifecycle @asyncapi/edavisualiser@1.2.0~prepare: Failed to exec prepare script
lerna ERR! lifecycle "prepare" errored in "@asyncapi/edavisualiser", exiting 1
npm verb exit 1
npm verb code 1
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/lib/index.js:53:27)
npm verb stack     at ChildProcess.emit (node:events:514:28)
npm verb stack     at maybeClose (node:internal/child_process:1091:16)
npm verb stack     at ChildProcess._handle.onexit (node:internal/child_process:302:5)
npm verb cwd /home/ankur/Desktop/development/EDAVisualiser
npm verb Linux 6.2.0-34-generic
npm verb node v18.17.1
npm verb npm  v9.6.7
npm ERR! code 1
npm ERR! path /home/ankur/Desktop/development/EDAVisualiser
npm ERR! command failed
npm ERR! command sh -c npm run bootstrap
npm verb exit 1
npm verb code 1

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-10-17T17_33_33_081Z-debug-0.log
```

**But when I ran the command the with the `sudo` privilege. It installed the package with some vulnerabilities**
```bash
up to date, audited 1157 packages in 25s

82 packages are looking for funding
  run `npm fund` for details

52 vulnerabilities (9 moderate, 38 high, 5 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force
```

**P.S.:**
**My node version is**
```bash
$ node --version
v18.17.1
```
**My npm version**
```bash
npm --version
9.6.7
```

**My Operating System**
**Ubuntu 22.04.3 LTS**

derberg commented:
yeah, not sure how to fix it, I'm not codeowners, so @jonaslagoni must decide how to fix dependency issues (if just bumping webpack is enough) or maybe you should switch to node 16 and package lock v2 🤷🏼

jonaslagoni commented:
We can bump everything to v18 😄 Dont think any of the implementation need to change (or dependant libraries) 🤔

ankur0904 commented:
@derberg @jonaslagoni Does this PR require any additional changes?

derberg commented:
@ankur0904 maybe try to update `webpack` in `library` dir to `5.89.0`. Might be you will have to update other webpack-related dependencies to make sure node v18 works

ankur0904 commented:
@derberg 
I have run the `test` in my local system and it passed
```bash
$ npm test --if-present

> test
> cd library && npm test


> @asyncapi/edavisualiser@1.2.0 test
> npm run test:unit


> @asyncapi/edavisualiser@1.2.0 test:unit
> jest --env=jsdom --detectOpenHandles .

 PASS  test/helpers/random-background-colors.spec.ts
  random-background-color
    ✓ should generate the same output when the same input hash is used (3 ms)
    ✓ should generate the different output when different inputs are used

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        1.058 s
Ran all test suites matching /./i.
```

ankur0904 commented:
@derberg I changed the version of `webpack` to `5.89.0` and deleted `node_modules` as well as `package-lock.json` from both place one from the main project and from the `library` folder also. But when I tried to install by `npm install --loglevel verbose` 
**ERROR**: 
```bash
lerna ERR! npm install exited 1 in '@asyncapi/edavisualiser'
lerna ERR! npm install stderr:
npm verb cli /home/ankur/.nvm/versions/node/v18.18.2/bin/node /home/ankur/.nvm/versions/node/v18.18.2/bin/npm
npm info using npm@9.8.1
npm info using node@v18.18.2
npm verb title npm install
npm verb argv "install"
npm verb logfile logs-max:10 dir:/home/ankur/.npm/_logs/2023-11-07T08_28_50_237Z-
npm verb logfile /home/ankur/.npm/_logs/2023-11-07T08_28_50_237Z-debug-0.log
npm http fetch GET 200 https://registry.npmjs.org/@asyncapi%2fparser 1980ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/canvas 1189ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/dompurify 1203ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/highlight.js 361ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/marked 66ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/react-flow-renderer 1158ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/react 82ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/react-dom 82ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/tailwindcss 71ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/autoprefixer 1214ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/postcss 67ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@cypress%2fwebpack-preprocessor 2583ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@babel%2fcore 69ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@babel%2fpreset-env 88ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/babel-loader 70ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/webpack 72ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@tailwindcss%2ftypography 2627ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fdompurify 2381ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fenzyme 3179ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fhighlight.js 1741ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fjest 2535ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fjsdom 1495ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fmarked 2379ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2fnode 1487ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2freact 89ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2freact-dom 73ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/@types%2freact-test-renderer 1666ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/cross-env 429ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/cssnano 60ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/cypress 1230ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/enzyme 1167ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/enzyme-adapter-react-16 366ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/jest 1225ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/postcss-cli 1208ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/postcss-import 55ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/postcss-scopify 1211ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/react-test-renderer 489ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/ts-jest 1230ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/typescript 92ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/ts-loader 1204ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/webpack-bundle-analyzer 1204ms (cache revalidated)
npm http fetch GET 200 https://registry.npmjs.org/webpack-cli 1182ms (cache revalidated)
npm verb stack Error: unable to resolve dependency tree
npm verb stack     at #failPeerConflict (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:1376:25)
npm verb stack     at #loadPeerSet (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:1342:33)
npm verb stack     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
npm verb stack     at async #buildDepStep (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:917:11)
npm verb stack     at async Arborist.buildIdealTree (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:196:7)
npm verb stack     at async Promise.all (index 1)
npm verb stack     at async Arborist.reify (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:159:5)
npm verb stack     at async Install.exec (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/lib/commands/install.js:149:5)
npm verb stack     at async module.exports (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/lib/cli-entry.js:61:5)
npm verb cwd /home/ankur/Desktop/development/EDAVisualiser/library
npm verb Linux 6.2.0-36-generic
npm verb node v18.18.2
npm verb npm  v9.8.1
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: @asyncapi/edavisualiser@1.2.0
npm ERR! Found: webpack@5.89.0
npm ERR! node_modules/webpack
npm ERR!   dev webpack@"^5.89.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer webpack@"4.x.x" from webpack-cli@3.3.12
npm ERR! node_modules/webpack-cli
npm ERR!   dev webpack-cli@"^3.3.12" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! /home/ankur/.npm/_logs/2023-11-07T08_28_50_237Z-eresolve-report.txt
npm verb exit 1
npm verb unfinished npm timer reify 1699345730402
npm verb unfinished npm timer reify:loadTrees 1699345730406
npm verb unfinished npm timer idealTree:buildDeps 1699345730418
npm verb unfinished npm timer idealTree:#root 1699345730418
npm verb code 1

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T08_28_50_237Z-debug-0.log

lerna ERR! npm install exited 1 in '@asyncapi/edavisualiser'
npm verb exit 1
npm verb code 1
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/lib/index.js:53:27)
npm verb stack     at ChildProcess.emit (node:events:517:28)
npm verb stack     at maybeClose (node:internal/child_process:1098:16)
npm verb stack     at ChildProcess._handle.onexit (node:internal/child_process:303:5)
npm verb cwd /home/ankur/Desktop/development/EDAVisualiser
npm verb Linux 6.2.0-36-generic
npm verb node v18.18.2
npm verb npm  v9.8.1
npm ERR! code 1
npm ERR! path /home/ankur/Desktop/development/EDAVisualiser
npm ERR! command failed
npm ERR! command sh -c npm run bootstrap
npm verb exit 1
npm verb code 1

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T08_28_15_124Z-debug-0.log
```

After I tried `npm install --force`
**ERROR**
```bash
npm ERR! node-pre-gyp info it worked if it ends with ok
npm ERR! node-pre-gyp info using node-pre-gyp@1.0.11
npm ERR! node-pre-gyp info using node@18.18.2 | linux | x64
npm ERR! node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz
npm ERR! node-pre-gyp ERR! install request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz failed, reason: connect ETIMEDOUT 64:ff9b::14cf:4952:443 
npm ERR! node-pre-gyp WARN Pre-built binaries not installable for canvas@2.11.2 and node@18.18.2 (node-v108 ABI, glibc) (falling back to source compile with node-gyp) 
npm ERR! node-pre-gyp WARN Hit error request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz failed, reason: connect ETIMEDOUT 64:ff9b::14cf:4952:443 
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@18.18.2 | linux | x64
npm ERR! gyp info ok 
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@18.18.2 | linux | x64
npm ERR! gyp info find Python using Python version 3.10.12 found at "/usr/bin/python3"
npm ERR! gyp info spawn /usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/.cache/node-gyp/18.18.2/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/ankur/.cache/node-gyp/18.18.2',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/ankur/.cache/node-gyp/18.18.2/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! Package cairo was not found in the pkg-config search path.
npm ERR! Perhaps you should add the directory containing `cairo.pc'
npm ERR! to the PKG_CONFIG_PATH environment variable
npm ERR! No package 'cairo' found
npm ERR! gyp: Call to 'pkg-config cairo --libs' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:325:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:517:28)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:292:12)
npm ERR! gyp ERR! System Linux 6.2.0-36-generic
npm ERR! gyp ERR! command "/home/ankur/.nvm/versions/node/v18.18.2/bin/node" "/home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--update-binary" "--module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node" "--module_name=canvas" "--module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v108"
npm ERR! gyp ERR! cwd /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas
npm ERR! gyp ERR! node -v v18.18.2
npm ERR! gyp ERR! node-gyp -v v9.4.0
npm ERR! gyp ERR! not ok 
npm ERR! node-pre-gyp ERR! build error 
npm ERR! node-pre-gyp ERR! stack Error: Failed to execute '/home/ankur/.nvm/versions/node/v18.18.2/bin/node /home/ankur/.nvm/versions/node/v18.18.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --update-binary --module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release --napi_version=9 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v108' (1)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:517:28)
npm ERR! node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1098:16)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:303:5)
npm ERR! node-pre-gyp ERR! System Linux 6.2.0-36-generic
npm ERR! node-pre-gyp ERR! command "/home/ankur/.nvm/versions/node/v18.18.2/bin/node" "/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build" "--update-binary"
npm ERR! node-pre-gyp ERR! cwd /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas
npm ERR! node-pre-gyp ERR! node -v v18.18.2
npm ERR! node-pre-gyp ERR! node-pre-gyp -v v1.0.11
npm ERR! node-pre-gyp ERR! not ok

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T08_35_23_762Z-debug-0.log

lerna ERR! npm install exited 1 in '@asyncapi/edavisualiser'
npm ERR! code 1
npm ERR! path /home/ankur/Desktop/development/EDAVisualiser
npm ERR! command failed
npm ERR! command sh -c npm run bootstrap

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T08_34_36_647Z-debug-0.log
```

`NodeJS Version:` v18.18.2
`npm version:` 9.8.1

derberg commented:
you also need to bump `webpack-cli` as current version used requires `webpack` v4. You need `webpack-cli` that requires `webpack` v5

ankur0904 commented:
@derberg Still have error.

derberg commented:
@ankur0904 you did not push anything, so you need to copy paste error for me, otherwise I won't be able to help

ankur0904 commented:
@derberg Sorry for that, I missed out.

ankur0904 commented:
```bash
npm http fetch POST 200 https://registry.npmjs.org/-/npm/v1/security/advisories/bulk 1805ms
npm WARN deprecated @types/highlight.js@10.1.0: This is a stub types definition. highlight.js provides its own type definitions, so you do not need this installed.
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm WARN deprecated sane@4.1.0: some dependency vulnerabilities fixed, support for node < 10 dropped, and newer ECMAScript syntax/features added
npm http fetch GET 200 https://registry.npmjs.org/@types/prop-types/-/prop-types-15.7.10.tgz 2303ms (cache miss)
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
npm WARN deprecated react-flow-renderer@9.6.9: react-flow-renderer has been renamed to reactflow, please use this package from now on https://reactflow.dev/docs/guides/migrate-to-v11/
npm info run canvas@2.11.2 install node_modules/canvas node-pre-gyp install --fallback-to-build --update-binary
npm info run canvas@2.11.2 install { code: 1, signal: null }
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/lib/index.js:53:27)
npm verb stack     at ChildProcess.emit (node:events:517:28)
npm verb stack     at maybeClose (node:internal/child_process:1098:16)
npm verb stack     at ChildProcess._handle.onexit (node:internal/child_process:303:5)
npm verb pkgid canvas@2.11.2
npm verb cwd /home/ankur/Desktop/development/EDAVisualiser/library
npm verb Linux 6.2.0-36-generic
npm verb node v18.18.0
npm verb npm  v9.8.1
npm ERR! code 1
npm ERR! path /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas
npm ERR! command failed
npm ERR! command sh -c node-pre-gyp install --fallback-to-build --update-binary
npm ERR! Failed to execute '/home/ankur/.nvm/versions/node/v18.18.0/bin/node /home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --update-binary --module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release --napi_version=9 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v108' (1)
npm ERR! node-pre-gyp info it worked if it ends with ok
npm ERR! node-pre-gyp verb cli [
npm ERR! node-pre-gyp verb cli   '/home/ankur/.nvm/versions/node/v18.18.0/bin/node',
npm ERR! node-pre-gyp verb cli   '/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/.bin/node-pre-gyp',
npm ERR! node-pre-gyp verb cli   'install',
npm ERR! node-pre-gyp verb cli   '--fallback-to-build',
npm ERR! node-pre-gyp verb cli   '--update-binary'
npm ERR! node-pre-gyp verb cli ]
npm ERR! node-pre-gyp info using node-pre-gyp@1.0.11
npm ERR! node-pre-gyp info using node@18.18.0 | linux | x64
npm ERR! node-pre-gyp verb command install []
npm ERR! node-pre-gyp http GET https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz
npm ERR! node-pre-gyp ERR! install request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz failed, reason: connect ETIMEDOUT 64:ff9b::14cf:4952:443 
npm ERR! node-pre-gyp WARN Pre-built binaries not installable for canvas@2.11.2 and node@18.18.0 (node-v108 ABI, glibc) (falling back to source compile with node-gyp) 
npm ERR! node-pre-gyp WARN Hit error request to https://github.com/Automattic/node-canvas/releases/download/v2.11.2/canvas-v2.11.2-node-v108-linux-glibc-x64.tar.gz failed, reason: connect ETIMEDOUT 64:ff9b::14cf:4952:443 
npm ERR! node-pre-gyp verb command build [ 'rebuild' ]
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp verb cli [
npm ERR! gyp verb cli   '/home/ankur/.nvm/versions/node/v18.18.0/bin/node',
npm ERR! gyp verb cli   '/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js',
npm ERR! gyp verb cli   'clean'
npm ERR! gyp verb cli ]
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@18.18.0 | linux | x64
npm ERR! gyp verb command clean []
npm ERR! gyp verb clean removing "build" directory
npm ERR! gyp info ok 
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp verb cli [
npm ERR! gyp verb cli   '/home/ankur/.nvm/versions/node/v18.18.0/bin/node',
npm ERR! gyp verb cli   '/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js',
npm ERR! gyp verb cli   'configure',
npm ERR! gyp verb cli   '--fallback-to-build',
npm ERR! gyp verb cli   '--update-binary',
npm ERR! gyp verb cli   '--module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node',
npm ERR! gyp verb cli   '--module_name=canvas',
npm ERR! gyp verb cli   '--module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release',
npm ERR! gyp verb cli   '--napi_version=9',
npm ERR! gyp verb cli   '--node_abi_napi=napi',
npm ERR! gyp verb cli   '--napi_build_version=0',
npm ERR! gyp verb cli   '--node_napi_label=node-v108'
npm ERR! gyp verb cli ]
npm ERR! gyp info using node-gyp@9.4.0
npm ERR! gyp info using node@18.18.0 | linux | x64
npm ERR! gyp verb command configure []
npm ERR! gyp verb find Python Python is not set from command line or npm configuration
npm ERR! gyp verb find Python Python is not set from environment variable PYTHON
npm ERR! gyp verb find Python checking if "python3" can be used
npm ERR! gyp verb find Python - executing "python3" to get executable path
npm ERR! gyp verb find Python - executable path is "/usr/bin/python3"
npm ERR! gyp verb find Python - executing "/usr/bin/python3" to get version
npm ERR! gyp verb find Python - version is "3.10.12"
npm ERR! gyp info find Python using Python version 3.10.12 found at "/usr/bin/python3"
npm ERR! gyp verb get node dir no --target version specified, falling back to host node version: 18.18.0
npm ERR! gyp verb command install [ '18.18.0' ]
npm ERR! gyp verb install input version string "18.18.0"
npm ERR! gyp verb install installing version: 18.18.0
npm ERR! gyp verb install --ensure was passed, so won't reinstall if already installed
npm ERR! gyp verb install version is already installed, need to check "installVersion"
npm ERR! gyp verb got "installVersion" 11
npm ERR! gyp verb needs "installVersion" 11
npm ERR! gyp verb install version is good
npm ERR! gyp verb get node dir target node version installed: 18.18.0
npm ERR! gyp verb build dir attempting to create "build" dir: /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build
npm ERR! gyp verb build dir "build" dir needed to be created? Yes
npm ERR! gyp verb python symlink creating symlink to "/usr/bin/python3" at "/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/node_gyp_bins/python3"
npm ERR! gyp verb build/config.gypi creating config file
npm ERR! gyp verb build/config.gypi writing out config file: /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/config.gypi
npm ERR! gyp verb config.gypi checking for gypi file: /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/config.gypi
npm ERR! gyp verb common.gypi checking for gypi file: /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/common.gypi
npm ERR! gyp verb gyp gyp format was not specified; forcing "make"
npm ERR! gyp info spawn /usr/bin/python3
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   '/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'make',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   '/home/ankur/.cache/node-gyp/18.18.0/include/node/common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/ankur/.cache/node-gyp/18.18.0',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/ankur/.cache/node-gyp/18.18.0/<(target_arch)/node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! Package cairo was not found in the pkg-config search path.
npm ERR! Perhaps you should add the directory containing `cairo.pc'
npm ERR! to the PKG_CONFIG_PATH environment variable
npm ERR! No package 'cairo' found
npm ERR! gyp: Call to 'pkg-config cairo --libs' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
npm ERR! gyp ERR! configure error 
npm ERR! gyp ERR! stack Error: `gyp` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onCpExit (/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:325:16)
npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:517:28)
npm ERR! gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:292:12)
npm ERR! gyp ERR! System Linux 6.2.0-36-generic
npm ERR! gyp ERR! command "/home/ankur/.nvm/versions/node/v18.18.0/bin/node" "/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--update-binary" "--module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node" "--module_name=canvas" "--module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release" "--napi_version=9" "--node_abi_napi=napi" "--napi_build_version=0" "--node_napi_label=node-v108"
npm ERR! gyp ERR! cwd /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas
npm ERR! gyp ERR! node -v v18.18.0
npm ERR! gyp ERR! node-gyp -v v9.4.0
npm ERR! gyp ERR! not ok 
npm ERR! node-pre-gyp ERR! build error 
npm ERR! node-pre-gyp ERR! stack Error: Failed to execute '/home/ankur/.nvm/versions/node/v18.18.0/bin/node /home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --update-binary --module=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release/canvas.node --module_name=canvas --module_path=/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas/build/Release --napi_version=9 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v108' (1)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/@mapbox/node-pre-gyp/lib/util/compile.js:89:23)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess.emit (node:events:517:28)
npm ERR! node-pre-gyp ERR! stack     at maybeClose (node:internal/child_process:1098:16)
npm ERR! node-pre-gyp ERR! stack     at ChildProcess._handle.onexit (node:internal/child_process:303:5)
npm ERR! node-pre-gyp ERR! System Linux 6.2.0-36-generic
npm ERR! node-pre-gyp ERR! command "/home/ankur/.nvm/versions/node/v18.18.0/bin/node" "/home/ankur/Desktop/development/EDAVisualiser/library/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build" "--update-binary"
npm ERR! node-pre-gyp ERR! cwd /home/ankur/Desktop/development/EDAVisualiser/library/node_modules/canvas
npm ERR! node-pre-gyp ERR! node -v v18.18.0
npm ERR! node-pre-gyp ERR! node-pre-gyp -v v1.0.11
npm ERR! node-pre-gyp ERR! not ok
npm verb exit 1
npm verb unfinished npm timer reify 1699365149512
npm verb unfinished npm timer reify:build 1699365323866
npm verb unfinished npm timer build 1699365323869
npm verb unfinished npm timer build:deps 1699365323869
npm verb unfinished npm timer build:run:install 1699365323900
npm verb unfinished npm timer build:run:install:node_modules/canvas 1699365323900
npm verb code 1

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T13_52_29_345Z-debug-0.log

lerna ERR! npm install exited 1 in '@asyncapi/edavisualiser'
npm verb exit 1
npm verb code 1
npm verb stack Error: command failed
npm verb stack     at ChildProcess.<anonymous> (/home/ankur/.nvm/versions/node/v18.18.0/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/lib/index.js:53:27)
npm verb stack     at ChildProcess.emit (node:events:517:28)
npm verb stack     at maybeClose (node:internal/child_process:1098:16)
npm verb stack     at ChildProcess._handle.onexit (node:internal/child_process:303:5)
npm verb cwd /home/ankur/Desktop/development/EDAVisualiser
npm verb Linux 6.2.0-36-generic
npm verb node v18.18.0
npm verb npm  v9.8.1
npm ERR! code 1
npm ERR! path /home/ankur/Desktop/development/EDAVisualiser
npm ERR! command failed
npm ERR! command sh -c npm run bootstrap
npm verb exit 1
npm verb code 1

npm ERR! A complete log of this run can be found in: /home/ankur/.npm/_logs/2023-11-07T13_51_52_527Z-debug-0.log
```

ankur0904 commented:
I made these changes
```js
...
    "webpack": "^5.89.0",
    "webpack-bundle-analyzer": "^4.4.0",
    "webpack-cli": "^5.1.4"
...
```

jonaslagoni commented:
@ankur0904 try add jest to the root package 🤨 https://github.com/ankur0904/EDAVisualiser/blob/a167fd0cb39d1023fa8abf697154fdd4b254f92c/package.json#L31

ankur0904 commented:
@jonaslagoni Done :)

ankur0904 commented:
@derberg @jonaslagoni 
Finally we got the **green checkmark**.

jonaslagoni reviewed (approved):
Nice perseverance @ankur0904 🔥👌

ankur0904 commented:
@jonaslagoni Thanks
:)
